File: March 2026 GFG SOLUTION/March-06.py
class Solution:
    def minWindow(self, s, p):
        n, m = len(s), len(p)
        if n < m:
            return ""
        fp, fs = [0] * 256, [0] * 256
        for c in p:
            fp[ord(c)] += 1
        l = cnt = 0
        minLen, start = float('inf'), -1
        for r in range(n):
            fs[ord(s[r])] += 1
            if fp[ord(s[r])] and fs[ord(s[r])] <= fp[ord(s[r])]:
                cnt += 1
            while cnt == m:
                if r - l + 1 < minLen:
                    minLen = r - l + 1
                    start = l
                fs[ord(s[l])] -= 1
                if fp[ord(s[l])] and fs[ord(s[l])] < fp[ord(s[l])]:
                    cnt -= 1
                l += 1
        return "" if start == -1 else s[start:start + minLen]

File: March 2026 GFG SOLUTION/March-06.java
class Solution {
    public static String minWindow(String s, String p) {
        int n = s.length(), m = p.length();
        if (n < m) return "";
        int[] fp = new int[256], fs = new int[256];
        for (char c : p.toCharArray()) fp[c]++;
        int l = 0, cnt = 0, minLen = Integer.MAX_VALUE, start = -1;
        for (int r = 0; r < n; r++) {
            fs[s.charAt(r)]++;
            if (fp[s.charAt(r)] > 0 && fs[s.charAt(r)] <= fp[s.charAt(r)]) cnt++;
            while (cnt == m) {
                if (r - l + 1 < minLen) {
                    minLen = r - l + 1;
                    start = l;
                }
                fs[s.charAt(l)]--;
                if (fp[s.charAt(l)] > 0 && fs[s.charAt(l)] < fp[s.charAt(l)]) cnt--;
                l++;
            }
        }
        return start == -1 ? "" : s.substring(start, start + minLen);
    }
}
